File: harmony-host/shared/runtime/plugins/GomokuPlugin.ts
import type { GamePlugin, Player, PluginResult } from "../types";
import { ok, err } from "../types";

export class GomokuPlugin implements GamePlugin {
  readonly id = "gomoku";
  private size = 15;
  private board: number[][] = Array.from({ length: 15 }, () => new Array(15).fill(0));
  private blackId: string | null = null;
  private whiteId: string | null = null;
  private blackName = "黑方";
  private whiteName = "白方";
  private currentStone = 1;
  private winnerId: string | null = null;
  private winnerName: string | null = null;
  private draw = false;
  private moveCount = 0;
  private started = false;
  private lastMove: [number, number] | null = null;

  onStart(players: Player[], _hostId: string): PluginResult {
    if (players.length < 2) return err("五子棋需要 2 名玩家");
    this.reset();
    this.blackId = players[0].id; this.whiteId = players[1].id;
    this.blackName = players[0].name; this.whiteName = players[1].name;
    this.started = true;
    return ok({ system: `开局 · ${this.blackName} 执黑 vs ${this.whiteName} 执白 · 黑先` });
  }

  onAction(playerId: string, payload: any): PluginResult {
    if (!this.started) return err("对局未开始");
    if (this.winnerId || this.draw) return err("对局已结束");
    const expected = this.currentStone === 1 ? this.blackId : this.whiteId;
    if (playerId !== expected) return err("还没轮到你");
    if (payload.x == null || payload.y == null) return err("缺少坐标");
    const x = payload.x | 0, y = payload.y | 0;
    if (x < 0 || x >= this.size || y < 0 || y >= this.size) return err("坐标越界");
    if (this.board[y][x] !== 0) return err("此处已有棋子");
    this.board[y][x] = this.currentStone;
    this.lastMove = [x, y];
    this.moveCount++;
    const moverName = this.currentStone === 1 ? this.blackName : this.whiteName;
    if (this.checkWin(x, y, this.currentStone)) {
      this.winnerId = playerId; this.winnerName = moverName;
      return ok({ finished: true, system: `🏆 ${moverName} 五子连珠，胜出！` });
    }
    if (this.moveCount >= this.size * this.size) {
      this.draw = true;
      return ok({ finished: true, system: `满盘和棋 · ${this.blackName} 与 ${this.whiteName} 战平` });
    }
    this.currentStone = this.currentStone === 1 ? 2 : 1;
    return ok({});
  }

  snapshot(): any {
    return {
      board: this.board, size: this.size,
      blackId: this.blackId, whiteId: this.whiteId,
      blackName: this.blackName, whiteName: this.whiteName,
      currentStone: this.currentStone,
      currentPlayerId: this.currentStone === 1 ? this.blackId : this.whiteId,
      winnerId: this.winnerId, winnerName: this.winnerName,
      draw: this.draw, moveCount: this.moveCount,
      lastMove: this.lastMove
    };
  }

  isFinished(): boolean { return this.winnerId != null || this.draw; }
  reset(): void {
    this.board = Array.from({ length: 15 }, () => new Array(15).fill(0));
    this.blackId = null; this.whiteId = null; this.blackName = "黑方"; this.whiteName = "白方";
    this.currentStone = 1; this.winnerId = null; this.winnerName = null;
    this.draw = false; this.moveCount = 0; this.started = false; this.lastMove = null;
  }

  private checkWin(x: number, y: number, stone: number): boolean {
    const dirs = [[1, 0], [0, 1], [1, 1], [1, -1]];
    for (const [dx, dy] of dirs) {
      const count = 1 + this.countDir(x, y, dx, dy, stone) + this.countDir(x, y, -dx, -dy, stone);
      if (count >= 5) return true;
    }
    return false;
  }
  private countDir(x: number, y: number, dx: number, dy: number, stone: number): number {
    let cx = x + dx, cy = y + dy, n = 0;
    while (cx >= 0 && cx < this.size && cy >= 0 && cy < this.size && this.board[cy][cx] === stone) {
      n++; cx += dx; cy += dy;
    }
    return n;
  }
}

File: android-host/app/src/main/java/com/huddle/host/runtime/plugins/GomokuPlugin.kt
package com.huddle.host.runtime.plugins

import com.huddle.host.runtime.GamePlugin
import com.huddle.host.runtime.Player
import com.huddle.host.runtime.PluginResult
import org.json.JSONArray
import org.json.JSONObject

/**
 * Authoritative 15x15 Gomoku. Board cells: 0 empty, 1 black, 2 white.
 */
class GomokuPlugin : GamePlugin {
    override val id: String = "gomoku"

    private val size = 15
    private var board: Array<IntArray> = Array(size) { IntArray(size) }
    private var blackId: String? = null
    private var whiteId: String? = null
    private var blackName: String = "黑方"
    private var whiteName: String = "白方"
    private var currentStone: Int = 1
    private var winnerId: String? = null
    private var winnerName: String? = null
    private var draw: Boolean = false
    private var moveCount: Int = 0
    private var started: Boolean = false
    private var lastMove: IntArray? = null

    override fun onStart(players: List<Player>, hostId: String): PluginResult {
        if (players.size < 2) {
            return PluginResult.Err("五子棋需要 2 名玩家")
        }
        reset()
        blackId = players[0].id
        whiteId = players[1].id
        blackName = players[0].name
        whiteName = players[1].name
        started = true
        return PluginResult.Ok(
            system = "开局 · $blackName 执黑 vs $whiteName 执白 · 黑先"
        )
    }

    override fun onAction(playerId: String, payload: JSONObject): PluginResult {
        if (!started) return PluginResult.Err("对局未开始")
        if (winnerId != null || draw) return PluginResult.Err("对局已结束")

        val expected = if (currentStone == 1) blackId else whiteId
        if (playerId != expected) {
            return PluginResult.Err("还没轮到你")
        }

        if (!payload.has("x") || !payload.has("y")) {
            return PluginResult.Err("缺少坐标")
        }
        val x = payload.getInt("x")
        val y = payload.getInt("y")
        if (x !in 0 until size || y !in 0 until size) {
            return PluginResult.Err("坐标越界")
        }
        if (board[y][x] != 0) {
            return PluginResult.Err("此处已有棋子")
        }

        board[y][x] = currentStone
        lastMove = intArrayOf(x, y)
        moveCount++

        val moverName = if (currentStone == 1) blackName else whiteName
        if (checkWin(x, y, currentStone)) {
            winnerId = playerId
            winnerName = moverName
            return PluginResult.Ok(
                finished = true,
                system = "🏆 $moverName 五子连珠，胜出！"
            )
        }
        if (moveCount >= size * size) {
            draw = true
            return PluginResult.Ok(
                finished = true,
                system = "满盘和棋 · $blackName 与 $whiteName 战平"
            )
        }

        currentStone = if (currentStone == 1) 2 else 1
        return PluginResult.Ok()
    }

    override fun snapshot(): JSONObject {
        val rows = JSONArray()
        for (y in 0 until size) {
            val row = JSONArray()
            for (x in 0 until size) {
                row.put(board[y][x])
            }
            rows.put(row)
        }
        return JSONObject()
            .put("board", rows)
            .put("size", size)
            .put("blackId", blackId)
            .put("whiteId", whiteId)
            .put("blackName", blackName)
            .put("whiteName", whiteName)
            .put("currentStone", currentStone)
            .put("currentPlayerId", if (currentStone == 1) blackId else whiteId)
            .put("winnerId", winnerId)
            .put("winnerName", winnerName)
            .put("draw", draw)
            .put("moveCount", moveCount)
            .put("lastMove", lastMove?.let { JSONArray().put(it[0]).put(it[1]) } ?: JSONObject.NULL)
    }

    override fun isFinished(): Boolean = winnerId != null || draw

    override fun reset() {
        board = Array(size) { IntArray(size) }
        blackId = null
        whiteId = null
        blackName = "黑方"
        whiteName = "白方"
        currentStone = 1
        winnerId = null
        winnerName = null
        draw = false
        moveCount = 0
        started = false
        lastMove = null
    }

    private fun checkWin(x: Int, y: Int, stone: Int): Boolean {
        val dirs = arrayOf(
            intArrayOf(1, 0),
            intArrayOf(0, 1),
            intArrayOf(1, 1),
            intArrayOf(1, -1)
        )
        for (d in dirs) {
            val count = 1 +
                countDir(x, y, d[0], d[1], stone) +
                countDir(x, y, -d[0], -d[1], stone)
            if (count >= 5) return true
        }
        return false
    }

    private fun countDir(x: Int, y: Int, dx: Int, dy: Int, stone: Int): Int {
        var cx = x + dx
        var cy = y + dy
        var n = 0
        while (cx in 0 until size && cy in 0 until size && board[cy][cx] == stone) {
            n++
            cx += dx
            cy += dy
        }
        return n
    }
}
